src/backend/main.ts

```diff
@@ -85,41 +86,175 @@ function hash(input: string, pepper: string) {
     .digest('base64');
 }
 
-export async function storeTotalResponsesToS3() {
+export async function storeDataDumpsToS3() {
+  //
+  // Validations
+
   const db = process.env.ATHENA_DB_NAME;
   if (!db) throw new Error('Athena DB name missing from environment');
   const bucket = process.env.BUCKET_NAME_OPEN_DATA;
   if (!bucket) throw new Error('Open data bucket name missing from environment');
   const domain = process.env.DOMAIN_NAME_OPEN_DATA;
   if (!domain) throw new Error('Open data domain name missing from environment');
 
-  const queryResult = await athenaExpress.query({
-    sql: 'SELECT COUNT(*) as total_responses FROM responses',
-    db,
-  });
+  //
+  // Perform Athena queries in parallel
+
+  const [totalResponsesResult, postalCodeLevelDataResponse] = await Promise.all([
+    athenaExpress.query({ sql: totalResponsesQuery, db }),
+    athenaExpress.query({ sql: postalCodeLevelDataQuery, db }),
+  ]);
+
+  //
+  // Map query results into the JSON format we want to see in s3
 
   // TODO: Add model for this
-  const data = queryResult.Items[0];
+  const totalResponses = totalResponsesResult.Items[0];
+
+  const cityLevelData = await mapPostalCodeLevelTopostalCodeLevelData(postalCodeLevelDataResponse.Items, bucket);
+
+  //
+  // Push data to S3
+  await s3PutJsonHelper({
+    Bucket: bucket,
+    Key: 'total_responses.json',
+    // TODO: If the cache is same for all JSON, move this to the helper
+    CacheControl: 'max-age=15',
+    Body: {
+      meta: {
+        description:
+          'Total number of responses collected by the system thus far. This is the raw number before any filtering or abuse detection has been performed.',
+        generated: new Date().toISOString(),
+        link: `https://${domain}/total_responses.json`,
+      },
+      data: totalResponses,
+    },
+  });
+
+  await s3PutJsonHelper({
+    Bucket: bucket,
+    Key: 'city_level_general_results.json',
+    CacheControl: 'max-age=15',
+    Body: {
+      meta: {
+        description:
+          'Population and response count per each city in Finland, where the response count was higher than 25. All the form inputs are coded in city level. Population data from Tilastokeskus. This data is released for journalistic and scientific purposes.',
+        generated: new Date().toISOString(),
+        link: `https://${domain}/city_level_general_results.json`,
+      },
+      data: cityLevelData,
+    },
+  });
+}
+
+async function mapPostalCodeLevelTopostalCodeLevelData(postalCodeLevelData: any[], bucket: string) {
+  const postalCodeCityMappings = await s3GetJsonHelper({ Bucket: bucket, Key: 'postalcode_city_mappings.json' });
+
+  const populationPerCity = await s3GetJsonHelper({ Bucket: bucket, Key: 'population_per_city.json' });
```

Comment:
Suppose these could also be parallelized..? Though the gains are probably negligible.

Reply:
I think we might be talking milliseconds, though it makes sense. I would even consider parallelizing the whole pipeline for the 2 datasets though, even though I'm not sure about the failure-semantics of `Promise.all`. Meaning, if the other operation rejects in the middle, does it orphan the other that would happily continue? 🤷‍♂️

Reply:
All tracks within `all()` will finish, but their results are then ignored outside the `all()`. So depends on how you set it up!

But yes, not very important rn. :smile: